handle_exec_cmd():890 error

2024-06-11/01:28:04 PM [ERROR] [src/config_parser.c:handle_exec_cmd():890] execlp failed

How do i figure out what in config caused the issue? My config is as below

```
exec = ["polybar", "--reload", "main"]

border_width = 2
active_border_color = 0x83a598
normal_border_color = 0x30302f
window_gap = 10
virtual_desktops = 7
focus_follow_pointer = true


; run processes on some keys
bind = super + return -> run("alacritty")
bind = super + p -> run(["rofi", "-show", "drun"])

; kill the focused window
bind = super + x -> func(kill)

; switch to specific virtual desktops
bind = super + 1 -> func(switch_desktop:1)
bind = super + 2 -> func(switch_desktop:2)
bind = super + 3 -> func(switch_desktop:3)
bind = super + 4 -> func(switch_desktop:4)
bind = super + 5 -> func(switch_desktop:5)
bind = super + 6 -> func(switch_desktop:6)
bind = super + 7 -> func(switch_desktop:7)

; resize the focused window
bind = super + l -> func(resize:grow)
bind = super + h -> func(resize:shrink)

; toggle fullscreen mode
bind = super + f -> func(fullscreen)

; swap the focused window with its sibling
bind = super + s -> func(swap)

; cycle focus between windows
bind = super + up -> func(cycle_window:up)
bind = super + right -> func(cycle_window:right)
bind = super + left -> func(cycle_window:left)
bind = super + down -> func(cycle_window:down)

; cycle through virtual desktops
bind = super|shift + left -> func(cycle_desktop:left)
bind = super|shift + right -> func(cycle_desktop:right)

; transfer the focused window to another virtual desktop
bind = super|shift + 1 -> func(transfer_node:1)
bind = super|shift + 2 -> func(transfer_node:2)
bind = super|shift + 3 -> func(transfer_node:3)
bind = super|shift + 4 -> func(transfer_node:4)
bind = super|shift + 5 -> func(transfer_node:5)
bind = super|shift + 6 -> func(transfer_node:6)
bind = super|shift + 7 -> func(transfer_node:7)

; change the layout
bind = super|shift + m -> func(layout:master)
bind = super|shift + s -> func(layout:stack)
bind = super|shift + d -> func(layout:default)

; traverse the stack layout
bind = super|shift + k -> func(traverse:up)
bind = super|shift + j -> func(traverse:down)

; flip the orientation of the window
bind = super|shift + f -> func(flip)

; reload the configuration file
bind = super|shift + r -> func(reload_config)
```

Commenting Line 867 gives me error while compiling
```gcc -Wall -Wextra -Wshadow -Wunreachable-code -Wcast-align -Wuninitialized -c src/config_parser.c -o src/config_parser.o
src/config_parser.c:893:11: error: expected identifier or ‘(’ before ‘else’
  893 |         } else if (pid < 0) {
      |           ^~~~
src/config_parser.c:897:1: error: expected identifier or ‘(’ before ‘}’ token
  897 | }
      | ^
make: *** [Makefile:19: src/config_parser.o] Error 1```

--- a/README.md
+++ b/README.md
@@ -117,7 +117,14 @@ ZWM uses **binary space partitioning tree** ([BSP-tree](https://en.wikipedia.org
 </p>
 
 ## Installation
+### You can grab the executable from the [releases](https://github.com/Yazeed1s/zwm/releases) page, or build from source.
 ### Build from source
+#### Dependencies
+- gcc
+- libxcb
+- xcb-util
+- xcb-util-keysyms
+- xcb-util-wm (ewmh,icccm)
 ```
 git clone https://github.com/Yazeed1s/zwm.git
 cd zwm && sudo make install

--- a/src/config_parser.c
+++ b/src/config_parser.c
@@ -864,7 +864,7 @@ handle_exec_cmd(char *cmd)
 	pid_t pid = fork();
 
 	if (pid == 0) {
-		if (strchr(cmd, ',') != 0) {
+		if (strchr(cmd, ',')) {
 			trim(cmd, SQUARE_BRACKET);
 			int	   count = 0;
 			char **s	 = split_string(cmd, ',', &count);
@@ -1017,6 +1017,8 @@ int
 load_config(config_t *c)
 {
 	const char *filename = CONF_PATH;
-	return !file_exists(filename) ? write_default_config(filename, c)
-								  : parse_config(filename, c, false);
+	if (!file_exists(filename)) {
+		write_default_config(filename, c);
+	}
+	return parse_config(filename, c, false);
 }

--- a/src/logger.c
+++ b/src/logger.c
@@ -34,6 +34,7 @@
 #include <xcb/xcb.h>
 
 #define YEAR_OFFSET 1900
+#define LOG_PATH	"~/.local/share/xorg/zwm.log"
 
 void
 log_message(log_level_t level, const char *format, ...)
@@ -46,7 +47,7 @@ log_message(log_level_t level, const char *format, ...)
 	t	= time(NULL);
 	ptr = localtime(&t);
 	strftime(buf, 100, "%F/%I:%M:%S %p", ptr);
-	FILE *log_file = fopen("zwm.log", "a");
+	FILE *log_file = fopen(LOG_PATH, "a");
 	va_start(args, format);
 	if (log_file == NULL) {
 		fprintf(stderr, "Failed to open log file for writing\n");

--- a/src/zwm.c
+++ b/src/zwm.c
@@ -79,6 +79,7 @@ uint16_t			  num_lock;
 uint16_t			  caps_lock;
 uint16_t			  scroll_lock;
 int8_t				  click_to_focus;
+bool				  is_kgrabbed = false;
 
 // clang-format off
 // X11/keysymdef.h
@@ -123,8 +124,8 @@ static const _key__t keys_[] = {
 };
 // clang-format on
 
-static const uint32_t buttons_[] = {
-	XCB_BUTTON_INDEX_1, XCB_BUTTON_INDEX_2, XCB_BUTTON_INDEX_3};
+static const uint32_t buttons_[]  = {
+	 XCB_BUTTON_INDEX_1, XCB_BUTTON_INDEX_2, XCB_BUTTON_INDEX_3};
 
 int
 len_of_digits(long n)
@@ -1667,6 +1668,7 @@ grab_keys(xcb_conn_t *conn, xcb_window_t win)
 				return -1;
 			}
 		}
+		is_kgrabbed = true;
 		return 0;
 	}
 
@@ -1693,7 +1695,7 @@ grab_keys(xcb_conn_t *conn, xcb_window_t win)
 			return -1;
 		}
 	}
-
+	is_kgrabbed = true;
 	return 0;
 }
 
@@ -2501,8 +2503,8 @@ handle_subsequent_window(client_t *client, desktop_t *d)
 		return 0;
 	}
 
-	node_t *n			= find_node_by_window_id(d->tree, wi);
-	bool	is_floating = false;
+	node_t *n = find_node_by_window_id(d->tree, wi);
+	// bool	is_floating = false;
 	if (n->client->state == FLOATING) {
 		_LOG_(INFO, "node under cusor is floating %d", wi);
 		n = find_left_leaf(d->tree);
@@ -3341,6 +3343,27 @@ handle_button_press_event(xcb_button_press_event_t *ev)
 	xcb_flush(wm->connection);
 }
 
+int
+handle_mapping_notify(xcb_mapping_notify_event_t *e)
+{
+	if (e->request != XCB_MAPPING_KEYBOARD &&
+		e->request != XCB_MAPPING_MODIFIER) {
+		return 0;
+	}
+
+	if (is_kgrabbed) {
+		ungrab_keys(wm->connection, wm->root_window);
+		is_kgrabbed = !is_kgrabbed;
+	}
+
+	if (0 != grab_keys(wm->connection, wm->root_window)) {
+		_LOG_(ERROR, "cannot grab keys");
+		return -1;
+	}
+
+	return 0;
+}
+
 void
 log_active_desktop(void)
 {
@@ -3541,12 +3564,9 @@ evet_loop(wm_t *w)
 			break;
 		}
 		case XCB_MAPPING_NOTIFY: {
-			__attribute__((unused))
 			xcb_mapping_notify_event_t *mapping_notify =
 				(xcb_mapping_notify_event_t *)event;
-			if (0 != grab_keys(wm->connection, wm->root_window)) {
-				_LOG_(ERROR, "cannot grab keys");
-			}
+			handle_mapping_notify(mapping_notify);
 		}
 		default: {
 			break;
@@ -3595,6 +3615,10 @@ main(int argc, char **argv)
 
 	if (argc >= 2) {
 		parse_args(argc, argv);
+	}
+
+	if (0 != grab_keys(wm->connection, wm->root_window)) {
+		_LOG_(ERROR, "cannot grab keys");
 	}
 
 	evet_loop(wm);

